rfcs/0010-additional-target-hooks.md

```diff
@@ -0,0 +1,155 @@
+```
+runtime::Module BuildWooflesHost(IRModule mod, Target target) {
+// ... Custom Code generation here
+}
+
+TVM_REGISTER_GLOBAL("target.build.woofles").set_body_typed(BuildWooflesHost);
+TVM_REGISTER_TARGET_KIND("woofles", kDLCPU)
+    .set_attr<String>("tir_to_runtime", "target.build.woofles");
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This functionality is an extension of the existing use of `attr::kCompiler` to provide a hint that we can use to lookup attached target attribute, the compile engine and code generation flows can choose to store TIR and/or generate runtime modules based on the registered hooks.
+
+## Relay to TIR Hook
+[relay-to-tir-hook]: #relay-to-tir-hook
+
+This can be added to the TE Compiler by cross referencing the existing `attr::kCompiler` with the `TargetKind` registry:
+```
+auto code_gen_name = key->source_func->GetAttr<String>(attr::kCompiler).value();
+auto target_kind = tvm::TargetKind::Get(code_gen_name).value();
+if (target_kind.defined()) {
+    auto map = tvm::TargetKind::GetAttrMap<String>("relay_to_tir");
+    std::string custom_lowering = map[target_kind];
+    auto lowering_function = tvm::runtime::Registry::Get(custom_lowering);
+    cache_node->target = key->target;
+    cache_node->funcs = (*lowering_function)(key->source_func, key->target);
+    return CachedFunc(cache_node);
+}
+```
+By placing this where lowering currently takes place, it means minimal changes to executor code generators as they call into `LowerTE` and thus are agnostic to how it gets lowered.
+
+## TIR to Runtime Hook
+[tir-to-runtime-hook]: #tir-to-runtime-hook
+Instead of replicating the current external code generation hook, it is proposed that this hook exists in `build_module.cc`:
+```
+auto target_built_mods = FindFuncsWithTargetBuild(lowered_funcs);
+auto ext_mods = executor_codegen_->GetExternalModules();
+auto extra_mods = ext_mods->Concat(target_built_mods);
+ret_.mod = tvm::codegen::CreateMetadataModule(ret_.params, ret_.mod, extra_mods, GetTargetHost(),
+                                                executor_codegen_->GetMetadata());
+```
+This means the hook is integrated at a higher level and included in the compile flow without executors having to be aware of how these modules exist. See [Relay to TIR Hook](#relay-to-tir-hook) for how the `TargetKind` registry would be used.
+
+## Relay to Runtime Hook
+[relay-to-runtime-hook]: #relay-to-runtime-hook
+This would replace the existing `relay.ext.<target>` lookup in `compile_engine.cc`, essentially using the same logic as [Relay to TIR Hook](#relay-to-tir-hook) to cross reference with `kCompiler`.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+* Different hooks are currently dealt with in quite disparate parts of the codebase which are being heavily refactored
```

Comment:
i would also say that introducing custom TIR has the potential to add edge cases to the compiler which may uncover new bugs.